python/GafferTest/NodeAlgoTest.py

```diff
@@ -187,11 +187,174 @@ def testPresetsArray( self ) :
 		self.assertEqual( node["op1"].getValue(), 10 )
 		self.assertEqual( Gaffer.NodeAlgo.currentPreset( node["op1"] ), "c" )
 
+	def __visitationGraph( self ) :
+
+		# L1_1     L1_2
+		#   |       |\
+		#   |       | \
+		#   |       |  \
+		# L2_1   L2_2   L2_3
+		#   |\      |   /
+		#   | \     |  /
+		#   |  \    | /
+		#   |   \   |/
+		# L3_1   L3_2
+
+		s = Gaffer.ScriptNode()
+
+		s["L1_1"] = GafferTest.MultiplyNode()
+		s["L1_2"] = GafferTest.AddNode()
+
+		s["L2_1"] = GafferTest.AddNode()
+		s["L2_2"] = GafferTest.MultiplyNode()
+		s["L2_3"] = GafferTest.AddNode()
+
+		s["L3_1"] = GafferTest.AddNode()
+		s["L3_2"] = GafferTest.MultiplyNode()
+		s["L3_2"]["op3"] = Gaffer.IntPlug()
+
+		s["L2_1"]["op1"].setInput( s["L1_1"]["product"] )
+		s["L2_2"]["op1"].setInput( s["L1_2"]["sum"] )
+		s["L2_3"]["op1"].setInput( s["L1_2"]["sum"] )
+
+		s["L3_1"]["op1"].setInput( s["L2_1"]["sum"] )
+		s["L3_2"]["op1"].setInput( s["L2_1"]["sum"] )
+		s["L3_2"]["op2"].setInput( s["L2_2"]["product"] )
+		s["L3_2"]["op3"].setInput( s["L2_3"]["sum"] )
+
+		return s
+
+	class __CapturingVisitor( object ) :
+
+		def __init__( self ) :
+
+			self.visited = []
+
+		def __call__( self, node ) :
+
+			self.visited.append( node )
+			return True
+
+	def testVisitUpstream( self ) :
+
+		g = self.__visitationGraph()
+
+		v = self.__CapturingVisitor()
+		Gaffer.NodeAlgo.visitUpstream( g["L3_1"], v )
+		self.assertEqual( v.visited, [ g["L2_1"], g["L1_1"] ] )
+
+		del v.visited[:]
+		Gaffer.NodeAlgo.visitUpstream( g["L3_2"], v )
+		self.assertEqual( v.visited, [ g["L2_1"], g["L2_2"], g["L2_3"], g["L1_1"], g["L1_2"] ] )
+
+		del v.visited[:]
+		Gaffer.NodeAlgo.visitUpstream( g["L3_2"], v, order = Gaffer.NodeAlgo.VisitOrder.DepthFirst )
+		self.assertEqual( v.visited, [ g["L2_1"], g["L1_1"], g["L2_2"], g["L1_2"], g["L2_3"] ] )
+
+	def testVisitDownstream( self ) :
+
+		g = self.__visitationGraph()
+
+		v = self.__CapturingVisitor()
+		Gaffer.NodeAlgo.visitDownstream( g["L1_1"], v )
+		self.assertEqual( v.visited, [ g["L2_1"], g["L3_1"], g["L3_2"] ] )
+
+		del v.visited[:]
+		Gaffer.NodeAlgo.visitDownstream( g["L1_2"], v )
+		self.assertEqual( v.visited, [ g["L2_2"], g["L2_3"], g["L3_2"] ] )
+
+		del v.visited[:]
+		Gaffer.NodeAlgo.visitDownstream( g["L1_2"], v, order = Gaffer.NodeAlgo.VisitOrder.DepthFirst )
+		self.assertEqual( v.visited, [ g["L2_2"], g["L3_2"], g["L2_3"] ] )
+
+	def testVisitConnected( self ) :
+
+		g = self.__visitationGraph()
+
+		v = self.__CapturingVisitor()
+		Gaffer.NodeAlgo.visitConnected( g["L2_1"], v )
+		self.assertEqual( v.visited, [ g["L1_1"], g["L3_1"], g["L3_2"], g["L2_2"], g["L2_3"], g["L1_2"] ] )
+
```

Comment:
Do we need a `DepthFirst` test for connected?

Reply:
Added in rebased d33d0eea73d99f25f8e4a82271c1de47069f6d85.